db/db_test.cc

```diff
@@ -4054,48 +4054,76 @@ TEST_F(DBTest, ConcurrentFlushWAL) {
   options.env = env_;
   WriteOptions wopt;
   ReadOptions ropt;
-  for (bool two_write_queues : {false, true}) {
-    for (bool manual_wal_flush : {false, true}) {
-      options.two_write_queues = two_write_queues;
-      options.manual_wal_flush = manual_wal_flush;
-      options.create_if_missing = true;
-      DestroyAndReopen(options);
-      std::vector<port::Thread> threads;
-      threads.emplace_back([&] {
-        for (size_t i = 0; i < cnt; i++) {
-          auto istr = ToString(i);
-          db_->Put(wopt, db_->DefaultColumnFamily(), "a" + istr, "b" + istr);
-        }
-      });
-      if (two_write_queues) {
-        threads.emplace_back([&] {
-          for (size_t i = cnt; i < 2 * cnt; i++) {
+  SyncPoint::GetInstance()->SetCallBack("SplitWrite:allowed", [&](void* arg) {
+    size_t& allowed = *(static_cast<size_t**>(arg))[0];
+    size_t size = *(static_cast<size_t**>(arg))[1];
+    // when 'allowed == size', it means we are going to start write the buffer,
+    // then we split write into two parts by allowed /= 2, then first write will
+    // be done immediately, and this callback will be invoked again in the next
+    // time with 'allowed != size', so we slow down the write to increase the
+    // chance of conflict between FlushWAL thread and Writebatch Put thread
+    if (allowed == size && size >= 4) {
+      allowed /= 2;
+    } else {
+      env_->SleepForMicroseconds(1000);  // 1ms
```

Comment:
It's not reliable to use `sleep` to cause certain interleaving between concurrent threads.
I may need to read the code more and understand this test better before giving a better suggestion that will use TEST_SYNC_POINTS.

Reply:
I just remove this call_back_para.
There is a simple way to reproduce a WAL corruption by doing concurrent write and FlushWAL. It is not necessary to reproduce another WAL corruption issue with a complicated code.